hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/store/CachingStateFrame.java

```diff
@@ -0,0 +1,203 @@
+            @NonNull final UpdatableReferenceCache<K> cache,
+            @NonNull final K key,
+            @NonNull final Object value);
+
+    /** Delete a value (of given type) from the given cache, given the key */
+    protected abstract void deleteValue(
+            @NonNull final Class<?> klass, @NonNull final UpdatableReferenceCache<K> cache, @NonNull final K key);
+
+    /** Return all the internal caches of this frame, holding all the different types of values */
+    @NonNull
+    protected Map<Class<?>, UpdatableReferenceCache<K>> getInternalCaches() {
+        return accessors.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, kv -> kv.getValue().cache));
+    }
+
+    /** Enables type-safe access to one type of cached entities in this level of the cache.  (Accessor is _tied_
+     * to this frame of the cache.) */
+    @SuppressWarnings("java:S3655") // "Optional value should only be accessed after checking presence"
+    //                              // - doesn't understand`flatMap`
+    protected class AccessorImpl<V> implements Accessor<K, V> {
+
+        @NonNull
+        private final Class<V> klass;
+
+        @NonNull
+        private final UpdatableReferenceCache<K> cache;
+
+        private AccessorImpl(@NonNull final Class<V> klass, @NonNull final UpdatableReferenceCache<K> cache) {
+            this.klass = klass;
+            this.cache = cache;
+        }
+
+        /** Get a value from this level of the cache */
+        @Override
+        public Optional<V> get(@NonNull final K key) {
+            final var oe = getValue(klass, cache, key);
+            try {
+                return oe.flatMap(o -> Optional.of(klass.cast(o)));
+            } catch (final ClassCastException ex) {
+                throw new CacheAccessIncorrectType(
+                        "Accessor for class %s fetched object of class %s"
+                                .formatted(
+                                        klass.getTypeName(), oe.get().getClass().getTypeName()),
+                        ex);
+            }
+        }
+
+        /** Set a new value for a value (in this level of the cache) */
+        @Override
+        public void set(@NonNull final K key, @NonNull V value) {
+            if (!klass.isInstance(value)) {
+                throw new CacheAccessIncorrectType("Trying to store %s in accessor for class %s"
+                        .formatted(value.getClass().getTypeName(), klass.getTypeName()));
+            }
+            setValue(klass, cache, key, value);
+        }
+
+        /** Delete a value (from in this level of the cache) */
+        @Override
+        public void delete(@NonNull final K key) {
+            // Can't check type of value matches because we don't have one
+            deleteValue(klass, cache, key);
+        }
+    }
+
+    /** Signals that a type error occurred with the _value_ type */
+    public static class CacheAccessIncorrectType extends RuntimeException {
```

Comment:
We should use the standard exception hierarchy.
```suggestion
    public static class CacheAccessIncorrectType extends EvmException {
```